### ISSUE TITLE
There is no nl_func_call_start[_single] option

### ISSUE THREAD

**Issue by arneforlie** [uncrustify/uncrustify#2020]
nl_func_call_empty = remove changed behaviour in commit 2fbe5a9 (Merge pull request #1819 from CDanU/fix-nl_func_decl_start)

I am trying to format the following so that in all cases the newline after opening ( is removed
```
class X21
{
public:
void f(
int p1, int p2);
};

void
X21::f(
int p1, int p2)
{
}

void
n1()
{
	X21 x21;
	x21.f(
1, 1);
}
```
With commit 2fbe5a9 (and later commits) I get the following:
```
class X21
{
public:
void f(int p1, int p2);
};

void
X21::f(int p1, int p2)
{
}

void
n1()
{
	X21 x21;
	x21.f(
		1, 1);
}
```

However, with commit cc930f5 (the one before) it works as I want it (the newline in the function call is removed):
```
class X21
{
public:
void f(int p1, int p2);
};

void
X21::f(int p1, int p2)
{
}

void
n1()
{
	X21 x21;
	x21.f(1, 1);
}
```

The config for all cases is:
```
nl_func_decl_start = remove
nl_func_def_start  = remove
nl_func_call_empty = remove
```

Based on the name, I'm not sure that nl_func_call_empty was supposed to work like this before, but it did.
What option can be used now to remove the newline when the function call is not empty?

**Comment by arneforlie** [comment]
ons. 5. sep. 2018 kl. 21:25 skrev Matthew Woehlke <notifications@github.com
>:

> What option can be used now to remove the newline when the function call
> is not empty?
>
> I don't see such an option. I guess it would be
> nl_func_call_start_multi_line if that was an I/A/R/F, but it isn't; it is
> effectively ignore/add only. (Several other options are in the same boat.)
>
> I think the change in behavior is correct. However, I'm also inclined to
> think there should be a way to remove the newline. Where it gets sticky is
> that you probably only want to remove it if doing so won't make the line
> "too long", which means the desired logic is rather more complicated than
> just changing nl_func_call_start_multi_line to be I/A/R/F.
>

The functionality already works for a method declaration and definition.
Why not use the same logic?

**Comment by mwoehlke-kitware** [comment]
> The functionality already works for a method declaration and definition.

Does it? Is uncrustify smart enough to remove the newline iff the arg list will "fit"?

Eh, well... I suppose there's an argument that we should have `nl_func_call_start` and `nl_func_call_start_single`, which should relate to `nl_func_call_start_multi_line` the same way the `_decl_` flavors interact. But I'd still consider "be smarter" an open feature request.

**Comment by arneforlie** [comment]
ons. 5. sep. 2018 kl. 22:01 skrev Matthew Woehlke <notifications@github.com
>:

> The functionality already works for a method declaration and definition.
>
> Does it? Is uncrustify smart enough to remove the newline iff the arg list
> will "fit"?
>
> I don’t know smart it is, but my point was that since uncrustify is able
to remove the newline for those to constructs, why not use the same logic
for a function call?

Eh, well... I suppose there's an argument that we should have
> nl_func_call_start and nl_func_call_start_single, which should relate to
> nl_func_call_start_multi_line the same way the _decl_ flavors interact.
> But I'd still consider "be smarter" an open feature request.
>

Sure, smartness can come later.

**Comment by mwoehlke-kitware** [comment]
> But I'd still consider "be smarter" an open feature request.

Actually, thinking more about it, I guess I can live with not opportunistically removing "unnecessary" newlines... particularly as I recently encountered a case when I wouldn't want to:

```c++
if (cond)
  some_func(
    arg1, arg2);
else
  some_longer_func(
    arg1, arg2);
```

(Braces elided for brevity.) Both calls had the same argument list. The first would have fit on one line, but the second didn't; I decided it was "prettier" to make them consistent by breaking both.

...so I don't think I'd use "smart" newline removal anyway.

(This does not contradict my claim that "there's an argument that we should have `nl_func_call_start` and `nl_func_call_start_single`"; note that I have updated the summary accordingly.)
